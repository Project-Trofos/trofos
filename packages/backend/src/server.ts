import express from 'express';
import cors from 'cors';
import cookieParser from 'cookie-parser';
import accountRouter from './routes/account.route';
import courseRouter from './routes/course.route';
import projectRouter from './routes/project.route';
import backlogRouter from './routes/backlog.route';
import userRouter from './routes/user.route';
import sprintRouter from './routes/sprint.route';
import feedbackRouter from './routes/feedback.route';
import roleRouter from './routes/role.route';
import githubRouter from './routes/github.route';
import settingsRouter from './routes/settings.route';

const app = express();

export const port = 3001;
export const frontendUrl = process.env.FRONTEND_BASE_URL || 'http://localhost:3000';

export const corsOptions = {
  origin: frontendUrl,
  credentials: true,
};

app.use(express.json());
app.use(cookieParser());
app.use(cors(corsOptions));

app.get('/', (req: express.Request, res: express.Response) => {
  res.send('Hello World!');
});

const router = express.Router()

router.use('/account', accountRouter);

// Routes for course
router.use('/course', courseRouter);

// Routes for project
router.use('/project', projectRouter);

// Routes for backlog
router.use('/backlog', backlogRouter);

// Routes for user
router.use('/user', userRouter);

// Routes for backlog
router.use('/sprint', sprintRouter);

// Routes for settings
router.use('/feedback', feedbackRouter);

// Routes for role
router.use('/role', roleRouter);

// Routes for github app
router.use('/github', githubRouter);

// Routes for settings
router.use('/settings', settingsRouter);

app.use('/api', router)
// For unit testing
export default app;
