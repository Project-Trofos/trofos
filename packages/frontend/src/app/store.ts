import { configureStore } from '@reduxjs/toolkit';
import { setupListeners } from '@reduxjs/toolkit/query';
import trofosApiSlice, { nusmodsApiSlice } from '../api';

const store = configureStore({
  reducer: {
    [trofosApiSlice.reducerPath]: trofosApiSlice.reducer,
    [nusmodsApiSlice.reducerPath]: nusmodsApiSlice.reducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(trofosApiSlice.middleware).concat(nusmodsApiSlice.middleware),
});

// optional, but required for refetchOnFocus/refetchOnReconnect behaviors
// see `setupListeners` docs - takes an optional callback as the 2nd arg for customization
setupListeners(store.dispatch);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export default store;
