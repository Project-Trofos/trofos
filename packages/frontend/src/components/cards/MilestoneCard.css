.common-card .milestone-card-range-picker {
  margin-bottom: 5px;
}
