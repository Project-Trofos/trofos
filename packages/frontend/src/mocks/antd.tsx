import React from 'react';
import { Select as AntdSelect, DatePicker as AntdDatePicker } from 'antd';
import mockDayjs from 'dayjs';

// Mock certain antd component to simplify testing in React Testing Library
vi.mock('antd', async () => { 
  const antd: any = await vi.importActual('antd');
  antd.theme.defaultConfig.hashed = false;

  function DatePicker(props: React.ComponentProps<typeof AntdDatePicker>) {
    const { defaultValue, placeholder, disabled, id, onChange } = props;
    return (
      <input
        id={id}
        defaultValue={defaultValue as React.SelectHTMLAttributes<HTMLInputElement>['defaultValue']}
        placeholder={placeholder}
        disabled={disabled}
        onChange={(e) => {
          if (onChange) {
            onChange(mockDayjs(e.target.value) as unknown as Parameters<typeof onChange>['0'], e.target.value);
          }
        }}
      />
    );
  }

  function RangePicker(props: React.ComponentProps<typeof AntdDatePicker.RangePicker>) {
    const { defaultValue, id, onChange } = props;
    return (
      <input
        id={id as string | undefined}
        defaultValue={defaultValue as React.SelectHTMLAttributes<HTMLInputElement>['defaultValue']}
        onChange={(e) => {
          if (onChange) {
            onChange(
              [mockDayjs(e.target.value), mockDayjs(e.target.value)] as unknown as Parameters<typeof onChange>['0'],
              [e.target.value, e.target.value],
            );
          }
        }}
      />
    );
  }

  DatePicker.RangePicker = RangePicker;

  function Select(props: React.ComponentProps<typeof AntdSelect>) {
    const { mode, value, defaultValue, className, onChange, disabled, children, id } = props;

    const multiple = ['tags', 'multiple'].includes(mode ?? '');

    return (
      <select
        id={id}
        value={value as React.SelectHTMLAttributes<HTMLSelectElement>['value']}
        defaultValue={defaultValue as React.SelectHTMLAttributes<HTMLSelectElement>['defaultValue']}
        multiple={multiple}
        disabled={disabled}
        className={className}
        onChange={(e) => {
          if (onChange) {
            onChange(
              multiple ? Array.from(e.target.selectedOptions).map((option) => option.value) : e.target.value,
              {},
            );
          }
        }}
      >
        {children}
      </select>
    );
  }

  function Option({ children, ...otherProps }: { children: React.ReactNode }) {
    return <option {...otherProps}>{children}</option>;
  }
  function OptGroup({ children, ...otherProps }: { children: React.ReactNode }) {
    return <optgroup {...otherProps}>{children}</optgroup>;
  }

  Select.Option = Option;
  Select.OptGroup = OptGroup;

  return { ...antd, Select, DatePicker };
});
