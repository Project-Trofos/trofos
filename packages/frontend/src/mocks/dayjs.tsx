const dayjsMock = {
  year: () => '2022',
};

export default dayjsMock;
