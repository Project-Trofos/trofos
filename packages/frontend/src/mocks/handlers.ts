/* eslint-disable import/no-extraneous-dependencies */
import { rest } from 'msw';
import { CourseData } from '../api/types';

const BASE_URL = process.env.REACT_APP_BACKEND_BASE_URL || 'http://localhost:3001';
const NUSMODS_URL = 'https://api.nusmods.com/v2/2022-2023/moduleList.json';

export const MSW_COURSE: CourseData = {
  id: 1,
  code: 'CS3203',
  cname: 'Software Engineering Project',
  startYear: 2022,
  endYear: 2022,
  startSem: 1,
  endSem: 1,
  description: null,
  milestones: [
    {
      id: 1,
      name: 'Milestone 1',
      course_id: 1,
      start_date: '2023-12-31',
      deadline: '2023-01-09',
      created_at: '2022-12-18',
    },
    {
      id: 2,
      name: 'Milestone 2',
      course_id: 1,
      start_date: '2023-01-10',
      deadline: '2023-01-19',
      created_at: '2022-12-18',
    },
    {
      id: 3,
      name: 'Milestone 3',
      course_id: 1,
      start_date: '2023-01-20',
      deadline: '2023-01-30',
      created_at: '2022-12-18',
    },
  ],
  users: [],
  public: false,
  created_at: '2022-09-14T03:33:34.960Z',
};

const PROJECT = {
  id: 1,
  pname: 'project1',
  pkey: null,
  description: 'project1_description',
  course_id: '1',
  public: false,
  created_at: '2022-09-15T01:58:01.735Z',
  course: MSW_COURSE,
};

const NUSMODS_MODULE = {
  moduleCode: 'EL1101E',
  title: 'The Nature of Language',
  semesters: [1, 2],
};

const handlers = [
  // Handles GET on /project
  rest.get(`${BASE_URL}/project/`, (req, res, ctx) => res(ctx.status(200), ctx.body(JSON.stringify([PROJECT])))),

  // Handles GET on /course
  rest.get(`${BASE_URL}/course/`, (req, res, ctx) => res(ctx.status(200), ctx.body(JSON.stringify([MSW_COURSE])))),

  // Handles GET to NUSMODS
  rest.get(NUSMODS_URL, (req, res, ctx) => res(ctx.status(200), ctx.body(JSON.stringify([NUSMODS_MODULE])))),

  // Handles POST on /project
  rest.post(`${BASE_URL}/project/`, (req, res, ctx) => res(ctx.status(200), ctx.body(JSON.stringify({})))),

  // Handles POST on /course
  rest.post(`${BASE_URL}/course/`, (req, res, ctx) => res(ctx.status(200), ctx.body(JSON.stringify({})))),

  // Handles POST on /course
  rest.post(`${BASE_URL}/course/project`, (req, res, ctx) => res(ctx.status(200), ctx.body(JSON.stringify({})))),

  rest.post(`${BASE_URL}/course/:courseId/milestone`, (req, res, ctx) =>
    res(ctx.status(200), ctx.body(JSON.stringify({}))),
  ),
];

export default handlers;
