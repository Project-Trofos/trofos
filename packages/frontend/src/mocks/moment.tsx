const momentMock = {
  year: () => '2022',
};

export default momentMock;
