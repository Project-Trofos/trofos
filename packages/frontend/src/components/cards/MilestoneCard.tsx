import React from 'react';
import { Button, Card, Input, Steps, DatePicker } from 'antd';
import { useCourse } from '../../api/hooks';
import { Subheading } from '../typography';
import MilestoneCreationModal from '../modals/MilestoneCreationModal';
import { useMilestone } from './useMilestone';
import { CourseData } from '../../api/types';

import './CommonCard.css';
import './MilestoneCard.css';

const { Step } = Steps;

type MilestoneCardProps = {
  course?: CourseData;
  showEdit?: boolean;
};

// Card for displaying and editing milestones of a course
export default function MilestoneCard(props: MilestoneCardProps): JSX.Element {
  const { course, showEdit = false } = props;
  const { handleDeleteMilestone, handleUpdateMilestone } = useCourse(course?.id.toString());
  const { milestones, statuses } = useMilestone(course?.milestones);

  return (
    <Card className="common-card">
      <div className="common-card-header">
        <Subheading>Milestones</Subheading>
        {course && showEdit && <MilestoneCreationModal courseId={course.id.toString()} />}
      </div>
      {milestones && milestones?.length > 0 ? (
        <Steps current={-1}>
          {milestones.map((milestone, index) => {
            return (
              <Step
                key={milestone.id}
                title={
                  <Title showEdit={showEdit} milestone={milestone} handleUpdateMilestone={handleUpdateMilestone} />
                }
                status={statuses[index]}
                description={
                  <Description
                    showEdit={showEdit}
                    milestone={milestone}
                    handleUpdateMilestone={handleUpdateMilestone}
                    handleDeleteMilestone={handleDeleteMilestone}
                  />
                }
              />
            );
          })}
        </Steps>
      ) : (
        <div>No milestones.</div>
      )}
    </Card>
  );
}

function Description(props: {
  showEdit: boolean;
  milestone: NonNullable<ReturnType<typeof useMilestone>['milestones']>[number];
  handleUpdateMilestone: ReturnType<typeof useCourse>['handleUpdateMilestone'];
  handleDeleteMilestone: (milestoneId: number) => Promise<void>;
}): JSX.Element {
  const { handleDeleteMilestone, handleUpdateMilestone, milestone, showEdit } = props;
  return showEdit ? (
    <div>
      <DatePicker.RangePicker
        onCalendarChange={(e) => {
          if (!e) {
            return;
          }
          handleUpdateMilestone(milestone.id, {
            milestoneName: undefined,
            milestoneStartDate: e[0] ?? undefined,
            milestoneDeadline: e[1] ?? undefined,
          });
        }}
        size="small"
        value={[milestone.start_date, milestone.deadline]}
        className="milestone-card-range-picker"
      />
      <Button danger size="small" onClick={() => handleDeleteMilestone(milestone.id)}>
        Delete
      </Button>
    </div>
  ) : (
    <div>
      <div>Start Date: {milestone.start_date.format('DD/MM/YYYY')}</div>
      <div>Deadline: {milestone.deadline.format('DD/MM/YYYY')}</div>
    </div>
  );
}

function Title(props: {
  showEdit: boolean;
  milestone: NonNullable<ReturnType<typeof useMilestone>['milestones']>[number];
  handleUpdateMilestone: ReturnType<typeof useCourse>['handleUpdateMilestone'];
}): JSX.Element {
  const { handleUpdateMilestone, milestone, showEdit } = props;
  if (!showEdit) {
    return <div>{milestone.name}</div>;
  }

  return (
    <Input
      key={milestone.id}
      onKeyDown={(e) => {
        if (e.key === 'Enter') {
          (e.target as typeof e.target & { blur: () => void }).blur();
        }
      }}
      onBlur={(e) => {
        if (e.target.value === milestone.name) {
          return;
        }
        handleUpdateMilestone(milestone.id, { milestoneName: e.target.value });
      }}
      size="small"
      defaultValue={milestone.name}
    />
  );
}
