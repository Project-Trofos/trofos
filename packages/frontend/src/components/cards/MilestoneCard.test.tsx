import React from 'react';
import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import '../../mocks/antd';

import store from '../../app/store';
import server from '../../mocks/server';
import MilestoneCard from './MilestoneCard';
import { MSW_COURSE } from '../../mocks/handlers';

describe('test course creation modal', () => {
  beforeAll(() => server.listen());
  afterEach(() => server.resetHandlers());
  afterAll(() => server.close());

  const setup = () => {
    const { baseElement, debug } = render(
      <Provider store={store}>
        <MilestoneCard courseId={MSW_COURSE.id.toString()} />
      </Provider>,
    );
    return { baseElement, debug };
  };

  describe('when rendering', () => {
    it('should render correct fields', async () => {
      const { baseElement } = setup();
      for (const milestone of MSW_COURSE.milestones) {
        // eslint-disable-next-line no-await-in-loop
        await screen.findByText(milestone.name);
      }

      expect(baseElement).toMatchSnapshot();
    });
  });
});
