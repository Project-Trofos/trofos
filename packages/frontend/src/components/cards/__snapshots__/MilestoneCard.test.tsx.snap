// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test milestone creation modal when rendering should render correct fields 1`] = `
<body>
  <div>
    <div
      class="ant-card ant-card-bordered milestone-card"
    >
      <div
        class="ant-card-body"
      >
        <div
          class="milestone-card-header"
        >
          <h4
            class="ant-typography"
          >
            Milestones
          </h4>
        </div>
        <div
          class="ant-steps ant-steps-horizontal ant-steps-label-horizontal"
        >
          <div
            class="ant-steps-item ant-steps-item-process"
          >
            <div
              class="ant-steps-item-container"
            >
              <div
                class="ant-steps-item-tail"
              />
              <div
                class="ant-steps-item-icon"
              >
                <span
                  class="ant-steps-icon"
                >
                  1
                </span>
              </div>
              <div
                class="ant-steps-item-content"
              >
                <div
                  class="ant-steps-item-title"
                >
                  <div>
                    Milestone 2
                  </div>
                </div>
                <div
                  class="ant-steps-item-description"
                >
                  <div>
                    <div>
                      Start Date: 
                      10/01/2023
                    </div>
                    <div>
                      Deadline: 
                      19/01/2023
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="ant-steps-item ant-steps-item-wait"
          >
            <div
              class="ant-steps-item-container"
            >
              <div
                class="ant-steps-item-tail"
              />
              <div
                class="ant-steps-item-icon"
              >
                <span
                  class="ant-steps-icon"
                >
                  2
                </span>
              </div>
              <div
                class="ant-steps-item-content"
              >
                <div
                  class="ant-steps-item-title"
                >
                  <div>
                    Milestone 3
                  </div>
                </div>
                <div
                  class="ant-steps-item-description"
                >
                  <div>
                    <div>
                      Start Date: 
                      20/01/2023
                    </div>
                    <div>
                      Deadline: 
                      30/01/2023
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="ant-steps-item ant-steps-item-finish"
          >
            <div
              class="ant-steps-item-container"
            >
              <div
                class="ant-steps-item-tail"
              />
              <div
                class="ant-steps-item-icon"
              >
                <span
                  class="ant-steps-icon"
                >
                  <span
                    aria-label="check"
                    class="anticon anticon-check ant-steps-finish-icon"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="check"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M912 190h-69.9c-9.8 0-19.1 4.5-25.1 12.2L404.7 724.5 207 474a32 32 0 00-25.1-12.2H112c-6.7 0-10.4 7.7-6.3 12.9l273.9 347c12.8 16.2 37.4 16.2 50.3 0l488.4-618.9c4.1-5.1.4-12.8-6.3-12.8z"
                      />
                    </svg>
                  </span>
                </span>
              </div>
              <div
                class="ant-steps-item-content"
              >
                <div
                  class="ant-steps-item-title"
                >
                  <div>
                    Milestone 1
                  </div>
                </div>
                <div
                  class="ant-steps-item-description"
                >
                  <div>
                    <div>
                      Start Date: 
                      31/12/2023
                    </div>
                    <div>
                      Deadline: 
                      09/01/2023
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
